services/src/main/java/com/linkedin/kmf/services/TopicManagementService.java

```diff
@@ -162,30 +163,46 @@ public void run() {
       try {
         TopicState topicState = topicState();
         if (topicState == null) {
+          if (_topicCreationEnabled) {
+            TopicFactory topicFactory = (TopicFactory) Class.forName(_topicFactoryConfig).getConstructor(Map.class).newInstance();
+            topicFactory.createTopicIfNotExist(_zkConnect, _topic, _configuredTopicReplicationFactor,
+              _partitionsToBrokerRatio, new Properties());
+            topicState = topicState();
+          } else {
+            throw new RuntimeException(_serviceName + ": topic, " + _topic + ", does not exist and topic creation is not enabled.");
+          }
+        }
+
+        int currentReplicationFactor = topicState.replicationFactor();
+        if (currentReplicationFactor == -1) {
+          LOG.info(_serviceName + " can't determine replication factor of monitored topic " + _topic + ".  Will try rebalance later.");
           return;
+        } else if (currentReplicationFactor >= 0 && currentReplicationFactor != _configuredTopicReplicationFactor) {
+          throw new RuntimeException(_serviceName + ": replication factor given as configuration "
+              + TopicManagementServiceConfig.TOPIC_REPLICATION_FACTOR_CONFIG + " does not match the topic's current replication factor.");
         }
+
         if (topicState.someBrokerWithoutLeader() || topicState.someBrokerMissingPartition() || topicState.insufficientPartitions()) {
 
           LOG.info(_serviceName + ": topic rebalance started.");
           LOG.debug(_serviceName + ": broker count " + topicState._allBrokers.size() + " partitionCount " + topicState._partitionInfo.size() + ".");
 
-          if (topicState.replicationFactor() == -1) {
-            LOG.info(_serviceName + " can't determine replication factor of monitored topic " + _topic + ".  Will try rebalance later.");
-            return;
-          }
           if (topicState.insufficientPartitions()) {
             int idealPartitionCount = (int) Math.ceil(_partitionsToBrokerRatio * topicState._allBrokers.size());
             int addPartitionCount = idealPartitionCount - topicState._partitionInfo.size();
             LOG.info(_serviceName + ": adding " + addPartitionCount + " partitions.");
             AdminUtils.addPartitions(_zkUtils, _topic, idealPartitionCount, null, false, RackAwareMode.Enforced$.MODULE$);
             topicState = topicState();
+            currentReplicationFactor = topicState.replicationFactor();
```

Comment:
run() has become a very long method and should probably be refactored into smaller methods like : createTopic and rebalance(), but that can be some other time as this PR has been out here for sometime and is already very out of sync with master.